src/runtime/stream.rs

```diff
@@ -0,0 +1,293 @@
+use std::{
+    net::{SocketAddr, ToSocketAddrs},
+    ops::DerefMut,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use tokio::io::{AsyncRead, AsyncWrite};
+
+use crate::{
+    cmap::conn::StreamOptions,
+    error::{ErrorKind, Result},
+    options::StreamAddress,
+};
+
+const DEFAULT_CONNECT_TIMEOUT: Duration = Duration::from_secs(10);
+
+/// A runtime-agnostic async stream possibly using TLS.
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug)]
+pub(crate) enum AsyncStream {
+    /// A basic TCP connection to the server.
+    Tcp(AsyncStreamInner),
+
+    /// A TLS connection over TCP.
+    Tls(tokio_rustls::client::TlsStream<AsyncStreamInner>),
+}
+
+/// A runtime-agnostic async stream.
+#[derive(Debug)]
+pub(crate) enum AsyncStreamInner {
+    /// Wrapper around `tokio::net:TcpStream`.
+    #[cfg(feature = "tokio-runtime")]
+    Tokio(tokio::net::TcpStream),
+
+    /// Wrapper around `async_std::net::TcpStream`.
+    #[cfg(feature = "async-std-runtime")]
+    AsyncStd(async_std::net::TcpStream),
+}
+
+#[async_trait::async_trait]
+trait AsyncConnect
+where
+    Self: Sized,
+{
+    async fn connect(address: &SocketAddr) -> Result<Self>;
+
+    async fn connect_timeout(address: &SocketAddr, timeout: Duration) -> Result<Self>;
+}
+
+#[cfg(feature = "tokio-runtime")]
+#[async_trait::async_trait]
+impl AsyncConnect for tokio::net::TcpStream {
+    async fn connect(address: &SocketAddr) -> Result<Self> {
+        let stream = Self::connect(address).await?;
+        Ok(stream)
+    }
+
+    async fn connect_timeout(address: &SocketAddr, timeout: Duration) -> Result<Self> {
+        let stream = Self::connect_timeout(address, timeout).await?;
+        Ok(stream)
+    }
+}
+
+#[cfg(feature = "async-std-runtime")]
+#[async_trait::async_trait]
+impl AsyncConnect for async_std::net::TcpStream {
+    async fn connect(address: &SocketAddr) -> Result<Self> {
+        let stream = Self::connect(address).await?;
+        Ok(stream)
+    }
+
+    async fn connect_timeout(address: &SocketAddr, timeout: Duration) -> Result<Self> {
+        let stream = Self::connect_timeout(address, timeout).await?;
+        Ok(stream)
+    }
+}
+
+async fn try_connect<C: AsyncConnect>(address: &SocketAddr, timeout: Duration) -> Result<C> {
```

Comment:
`try_connect` and `connect_stream` are copies of their sync equivalents, with the only changes being making them generic over the type that's connecting and making them async functions and awaiting where necessary.